src/EditorFeatures/Core/ValueTracking/ValueTrackingService.Visitor.cs

```diff
@@ -0,0 +1,266 @@
+            {
+                var collectorsAndArgumentMap = argumentOperations
+                    .Where(ShouldTrackArgument)
+                    // Clone the collector here to allow each argument to report multiple items.
+                    // See Clone() docs for more details
+                    .Select(argument => (collector: Clone(), argument))
+                    .ToImmutableArray();
+
+                var tasks = collectorsAndArgumentMap
+                    .Select(pair => Task.Run(() => pair.collector.VisitAsync(pair.argument, cancellationToken)));
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                var items = collectorsAndArgumentMap
+                    .Select(pair => pair.collector.ProgressCollector)
+                    .SelectMany(collector => collector.GetItems())
+                    .Reverse(); // ProgressCollector uses a Stack, and we want to maintain the order by arguments, so reverse
+
+                foreach (var item in items)
+                {
+                    ProgressCollector.Report(item);
+                }
+            }
+
+            /// <summary>
+            /// Clone the current collector into a new one with
+            /// the same parent but a separate progress collector.
+            /// This allows collection of items given the same state
+            /// as this collector while also keeping them "grouped" separately.
+            /// </summary>
+            /// <remarks>
+            /// This is useful for cases such as tracking arguments, where each
+            /// argument may be an expression or something else. We want to track each
+            /// argument expression in the correct order, but a single argument may produce
+            /// multiple items. By cloning we can track the items for each argument and then
+            /// gather them all at the end to report in the correct order.
+            /// </remarks>
+            private OperationCollector Clone()
+            {
+                var collector = new ValueTrackingProgressCollector
+                {
+                    Parent = ProgressCollector.Parent
+                };
+                return new OperationCollector(collector, Solution);
+            }
+
+            private static bool ShouldTrackArgument(IArgumentOperation argumentOperation)
+            {
+                // Ref or Out arguments always contribute data as "assignments"
+                // across method calls
+                return argumentOperation.Parameter?.IsRefOrOut() == true
+
+                    // If the argument value is an expression, binary operation, or
+                    // invocation then parts of the operation need to be evaluated
+                    // to see if they contribute data for value tracking
+                    || argumentOperation.Value is IExpressionStatementOperation
```

Comment:
so..  this is a really good location to *not* try to use expressions :)  trying to format and intesperse teh comments is very unweildy.  just do this:

```c#
                // Ref or Out arguments always contribute data as "assignments"
                // across method calls
                if (argumentOperation.Parameter?.IsRefOrOut() == true)
                    return true;
```

similarly, you can then do your other checks, which you can easily comment and isolate.

Dense expressions are nto a virtue :)

Reply:
🤷‍♂️ I kind of like it, but it's preference. I'd rather tailor to normality and readability for others. I'll change